### ISSUE TITLE
 7kaa Version 2.14.5 bufferoverflow and others

### ISSUE THREAD

**Issue by nemysis** [the3dfxdude/7kaa#48]
I'm Maintainer for openSUSE for Seven Kingdoms and music

https://build.opensuse.org/package/show/games/7kaa

https://build.opensuse.org/package/show/games/7kaa-music

I try to update Seven Kingdoms: Ancient Adversaries to 2.14.5 but get these build failures

```\n
[  129s] Preparing packages...
[  129s] 7kaa-2.14.5-0.x86_64
[  132s] ... running 01-check-debuginfo
[  132s] ... testing for empty debuginfo packages
[  132s] ... running 02-check-gcc-output
[  132s] ... testing for serious compiler warnings
[  132s]     (using /usr/lib/build/checks-data/check_gcc_output)
[  132s]     (using /var/tmp/build-root/openSUSE_Factory-x86_64/.build.log)
[  132s] 
[  132s] I: A function overflows or underflows an array access. This could be a real error,
[  132s]    but occasionaly this condition is also misdetected due to loop unrolling or strange pointer
[  132s]    handling. So this is warning only, please review.
[  132s] W: 7kaa arraysubscript OF_MARK.cpp:817
[  132s] 
[  132s] I: Program causes undefined operation
[  132s]    (likely same variable used twiceand post/pre incremented in the same expression).
[  132s]    e.g. x = x++; Split it in two operations.
[  132s] W: 7kaa sequence-point OVOLUME.cpp:73
[  132s] 
[  132s] I: Program is likely to break with new gcc. Try -fno-strict-aliasing.
[  132s] W: 7kaa strict-aliasing-punning OERRCTRL.cpp:286, 327, 357, 397, 483
[  132s] W: 7kaa strict-aliasing-punning OMP_CRC.cpp:98, 121, 164, 196, 219, 242, 265, 310, 343, 376, 399, 437, 460, 486, 525, 548, 571, 604, 636, 672, 707, 739, 762, 788, 811, 835, 856, 877
[  132s] W: 7kaa strict-aliasing-punning OREMOTE2.cpp:333, 367, 444, 460
[  132s] W: 7kaa strict-aliasing-punning OREMOTEM.cpp:345, 383, 480, 1760, 1762, 2701
[  132s] W: 7kaa strict-aliasing-punning OSYS.cpp:2450
[  132s] W: 7kaa strict-aliasing-punning OUNITI.cpp:200
[  132s] 
[  132s] I: Statement is overflowing a buffer
[  132s] E: 7kaa bufferoverflow /usr/include/bits/stdio2.h:65
```

Please correct this.

Which license is for 7kaa-music, look similar than MIT or BSD but is not same.

I have used in openSUSE, but this is not good

```
License:        SUSE-Freeware
```

http://choosealicense.com/

Please use one from

https://spdx.org/licenses/

http://license.opensuse.org/

**Comment by MicroVirus** [comment]
Thanks for the bug report. I looked at these messages and these are my findings:
1. *OF_MARK.cpp:* Inside `FirmMarket::clear_market_goods`. I tested what would happen if I cleared a market product and it indeed writes outside the buffer of  `market_product_array`, into `market_raw_array`, when clearing a non-raw product from the market. The cause is a typo (using `raw_id` instead of `product_raw_id`) inside the second clause.
2.  *OVOLUME.cpp:* Double assignment  of `rel_vol = rel_vol = ...` fixed to single assigment.
3. *strict-aliasing*: Will look at these later, as these errors are plentiful and potentially harder to fix.
4. *stdio2.h:* This is a file outside of 7kaa and presumably part of glibc (https://github.com/lattera/glibc/blob/master/libio/bits/stdio2.h). Ignored.

Pull request for these issues: #49

**Comment by nemysis** [comment]
Thanks this have fixed with OFIRMIF.cpp, OF_BASE.cpp, OF_CAMP.cpp, OF_MARK.cpp and OVOLUME.cpp.

Only rest are

```\n
[  144s] I: Program is likely to break with new gcc. Try -fno-strict-aliasing.
[  144s] W: 7kaa strict-aliasing-punning OERRCTRL.cpp:286, 327, 357, 397, 483
[  144s] W: 7kaa strict-aliasing-punning OMP_CRC.cpp:98, 121, 164, 196, 219, 242, 265, 310, 343, 376, 399, 437, 460, 486, 525, 548, 571, 604, 636, 672, 707, 739, 762, 788, 811, 835, 856, 877
[  144s] W: 7kaa strict-aliasing-punning OREMOTE2.cpp:333, 367, 444, 460
[  144s] W: 7kaa strict-aliasing-punning OREMOTEM.cpp:345, 383, 480, 1760, 1762, 2701
[  144s] W: 7kaa strict-aliasing-punning OSYS.cpp:2450
[  144s] W: 7kaa strict-aliasing-punning OUNITI.cpp:200
[  144s] 
[  144s] I: Statement is overflowing a buffer
[  144s] E: 7kaa bufferoverflow /usr/include/bits/stdio2.h:65
```

**Comment by the3dfxdude** [comment]
To answer your question on the license, the music is not freeware, MIT or BSD. The music remains copyrighted and the license is similar to CC BY-NC-ND (creative commons with no derivative works, non-commericial). The author of the music did not grant anything beyond freely sharing the music the game. And I couldn't get him to commit to a creative commons license, so I just wrote out the terms as he specified. The license is not going to be changed. This likely means you cannot ship it with SuSE, but if you want to ship it, you better consult with the SuSE legal team, and get it marked for a restricted repository.

Also please do not file issues based on log dumps from your build system (especially "warnings"). It is not our responsibility to make the game build on options / tool versions you choose. If you believe there is a bug, document the bug and file an issue on that.

I researched the report from what you sent a few weeks ago, and this is what I determined:

1) True there is a typo, but I think the code isn't factored right to begin with. Need to find my notes.
2) Likely never caused an issue. I think I tried disassembling it and couldn't see anything wrong. Probably a warning to help prevent other kinds of unintentional issues, as demonstrated by the example.
3) GCC 5.0 was brand new when you sent the original report. Only people that are on the bleeding edge (unstable distros) are affected. I think the change is probably controversial and not worth the effort. I'm not sure I am willing to take a fix on this, or even changing the command line options without being able to test that compiler on our end.
4) Please determine the reason for the message. I can agree, ignore.

**Comment by nemysis** [comment]
License) I see on https://spdx.org/licenses/ these CC BY-NC-ND License

```\n
CC-BY-NC-ND-1.0
CC-BY-NC-ND-2.0
CC-BY-NC-ND-2.5
CC-BY-NC-ND-3.0
CC-BY-NC-ND-4.0
```

Which is best to use with 7kaa for music?

1) Yes MicroVirus have all fixed in Code, what was bug for me.

2) Without these fixes in Code, I can't build in openSUSE at all your great Game.

3) Please look #49 I have now fixed for GCC5.

PS: I have ported 7kaa to openSUSE at user question.

**Comment by the3dfxdude** [comment]
As I said earlier, I cannot change the license. Please specify 'other', or 'non-free'. It is not by definition any of those you are listing.

Regarding 2) please be more specific about "can't build". Throwing '-Wall' is not something we reasonably expect to be a 'build'/'no-build' criteria, and likewise any other switch not provided by us. This is a choice by you or your distro maintainers, and not an appropriate one. I also think that openSuSE already can build/run this game without any changes, so any special requirements being imposed by your build system is a weak one needing more justification.

But please do not worry. MicroVirus pull request will be accepted when he gives the go ahead. We will continue to address the issues as appropriate.

**Comment by nemysis** [comment]
I have send and is accepted, please look

https://build.opensuse.org/package/show/games/7kaa

https://build.opensuse.org/package/show/games/7kaa-music

Later I change License to CC-BY-NC-ND-1.0, must write to openSUSE Developer which works with licenses.

**Comment by the3dfxdude** [comment]
nemysis said:
> Later I change License to CC-BY-NC-ND-1.0, must write to openSUSE Developer which works with licenses.

You can't change the license! Do not do this. The license for the music is not CC-BY-NC-ND!

**Comment by MicroVirus** [comment]
The current license file for 7kaa_music is wrong: https://build.opensuse.org/package/view_file/games/7kaa-music/7kaa-music.spec?expand=1
It says SUSE-Freeware. This needs to be set to the correct, custom license that the3dfxdude mentions. If this can't be done, then the package needs to be removed from the OpenSUSE repository.

**Comment by nemysis** [comment]
7kaa-music is not in normal openSUSE version, it is in Factory:NonFree, where are packages with not good licenses, and not found on

https://spdx.org/licenses/

https://build.opensuse.org/project/show/openSUSE:Factory:NonFree

https://build.opensuse.org/package/show/openSUSE:Factory:NonFree/7kaa-music

For these packages must be used licenses from

http://license.opensuse.org/

as Example

SUSE-Freeware

or other which is better for 7kaa-music

**Comment by MicroVirus** [comment]
Ah, right, Thanks for your response. I did not fully understand the situation. I'll let the3dfxdude decide on what is the correct way to handle the licensing here.

**Comment by the3dfxdude** [comment]
Well, I might be OK with SUSE-Redistributable-Content if that is a category label and not its own special license. Meaning that it does not impose its own terms of any kind. Please clarify the purpose.

**Comment by stoecker** [comment]
A little followup. The package was later rejected due an error I overlooked. You should really get rid of many of these warnings. Most of them usually indicate errors. A full list is here: https://build.opensuse.org/build/games/openSUSE_13.2/x86_64/7kaa/_log
I added a new patch, which fixes a buffer overflow and allows building now. The fix solves the issue, but it is still bad design. Such a function like sprintf should be part of the String class instead of assuming buffer size. At least there should be a size() function in String class returning the buffer size when used like here:
https://build.opensuse.org/package/view_file/games/7kaa/7kaa-2.14.5_overflow.patch?expand=1

**Comment by the3dfxdude** [comment]
stoecker,

Please file proper bug reports. We will only address them on a properly documented case-by-case basis. Turning on -Wall and saying "fix it" isn't going to fly here.

**Comment by the3dfxdude** [comment]
Unavowed,

Please bring in stoecker's MAX_STR_LEN patch into your pull request so we can address that in the same pull.

**Comment by MicroVirus** [comment]
Not unavowed, but I think you meant me :)

**Comment by stoecker** [comment]
> Please file proper bug reports. We will only address them on a properly documented case-by-case basis. Turning on -Wall and saying "fix it" isn't going to fly here.

I actually don't know this software at all or care about it. I only inform your about the facts in openSUSE packaging, as due the trouble I overlooked in the submission review of the package update I had to fix it myself. Whether or not you fix your code is nothing which bothers me.

**Comment by MicroVirus** [comment]
The issue is that many of these warnings are not indications of errors. We've reviewed the warning output of 7kaa plenty of times and most things are fine, so giving a compiler log and saying that there are warnings isn't directly useful. What is helpful is finding actual problems so we can fix them, so  thanks for finding and fixing the error with `snprintf`.

I'll probably skim through the build output you supplied to see if anything indicating a bug comes up and then investigate.

**Comment by the3dfxdude** [comment]
Is -Warray-bounds the switch giving us these possible out-of-bounds uses? If this discussion tracked down the remaining problems with that switch enabled, at this point we should go ahead and enable that one. Then we should close this issue on that, since the report was really for that mainly, and the "and others" part we ought not keep this open just for anything else that comes up.

**Comment by MicroVirus** [comment]
Ok, I checked the entire -Wall output list and I found 3 genuine uninitialised variable uses that might be able to occur, but none of these are serious. The rest of the warnings can all be ignored, barring any mistakes on my side as I checked them. Fixes upcoming in #49.
